fem/qfunction.hpp

```diff
@@ -0,0 +1,260 @@
+       global values.
+
+       Inside the Vector @a values, the index `i+vdim*j` corresponds to the
+       `i`-th vector component at the `j`-th quadrature point.
+    */
+   inline void GetValues(int idx, Vector &values) const;
+
+   /// Return the quadrature function values at an integration point.
+   /** The result is stored in the Vector @a values as a reference to the
+       global values. */
+   inline void GetValues(int idx, const int ip_num, Vector &values);
+
+   /// Return the quadrature function values at an integration point.
+   /** The result is stored in the Vector @a values as a copy to the
+       global values. */
+   inline void GetValues(int idx, const int ip_num, Vector &values) const;
+
+   /// Return all values associated with mesh element @a idx in a DenseMatrix.
+   /** The result is stored in the DenseMatrix @a values as a reference to the
+       global values.
+
+       Inside the DenseMatrix @a values, the `(i,j)` entry corresponds to the
+       `i`-th vector component at the `j`-th quadrature point.
+    */
+   inline void GetValues(int idx, DenseMatrix &values);
+
+   /// Return all values associated with mesh element @a idx in a const DenseMatrix.
+   /** The result is stored in the DenseMatrix @a values as a copy of the
+       global values.
+
+       Inside the DenseMatrix @a values, the `(i,j)` entry corresponds to the
+       `i`-th vector component at the `j`-th quadrature point.
+    */
+   inline void GetValues(int idx, DenseMatrix &values) const;
+
+   /// Get the IntegrationRule associated with entity (element or face) @a idx.
+   const IntegrationRule &GetIntRule(int idx) const
+   { return GetSpace()->GetIntRule(idx); }
+
+   /// Write the QuadratureFunction to the stream @a out.
+   void Save(std::ostream &out) const;
+
+   /// @brief Write the QuadratureFunction to @a out in VTU (ParaView) format.
+   ///
+   /// The data will be uncompressed if @a compression_level is zero, or if the
+   /// format is VTKFormat::ASCII. Otherwise, zlib compression will be used for
+   /// binary data.
+   void SaveVTU(std::ostream &out, VTKFormat format=VTKFormat::ASCII,
+                int compression_level=0) const;
+
+   /// @brief Save the QuadratureFunction to a VTU (ParaView) file.
+   ///
+   /// The extension ".vtu" will be appended to @a filename.
+   /// @sa SaveVTU(std::ostream &out, VTKFormat format=VTKFormat::ASCII,
+   ///             int compression_level=0)
+   void SaveVTU(const std::string &filename, VTKFormat format=VTKFormat::ASCII,
+                int compression_level=0) const;
+
+   virtual ~QuadratureFunction()
+   {
+      if (own_qspace) { delete qspace; }
+   }
+};
+
+// Inline methods
+
+inline void QuadratureFunction::GetValues(
+   int idx, Vector &values)
+{
+   const int s_offset = qspace->offsets[idx];
+   const int sl_size = qspace->offsets[idx+1] - s_offset;
+   values.NewDataAndSize(data + vdim*s_offset, vdim*sl_size);
```

Comment:
```suggestion
   values.NewDataAndSize(HostRead() + vdim*s_offset, vdim*sl_size);
```
?

Reply:
That looks right. This is just moving the implementation from `gridfunc.hpp` to a new file. @tzanio @v-dobrev, should it be `HostRead` here (and in other occurrences below)?

Reply:
Done in e0d8ee2bc. I used `MakeRef`.